File: agent-driven-architecture-main/examples/kotlin/spine/src/main/kotlin/adr/spine/boundary/Boundary.kt
// ── spine/boundary/boundary — THE ONE IMPURE SEAM (G9) ─────────────────────
// Nine ordered steps. Everything impure in the system happens here or in a block's
// single adapter; nothing else.
//
// Three structural facts fall out, and every implementation must preserve all three:
//
//  * COMMIT STRICTLY PRECEDES PERFORM — not by convention, but because step 9
//    cannot run until step 7 has returned the StepIndex the effect key is built
//    from. 14.6's ordering claim becomes unwritable-wrong.
//
//  * THE GATE RUNS BEFORE THE FOLD AND BEFORE THE COMMIT, so what is committed is
//    already the gate's verdict. A re-fold reproduces it without calling the
//    authorization seam again (G9).
//
//  * NOTHING DOWNSTREAM OF STEP 4 CAN LEARN WHO ACTED EXCEPT THROUGH `sig`. The
//    results were produced in step 3, before the signature existed. G1's
//    two-unreconciled-actor-values problem cannot recur, because there is only one
//    value and it is created after the tool has returned.
//
// AND NOTHING UPSTREAM OF STEP 4 CAN CHOOSE IT EITHER. `by` is a parameter of the
// CHANNEL a caller was handed, never a property of the step it submits, so the value
// fed to `authorityOf` is decided in this file and nowhere else.

package adr.spine.boundary

import adr.spine.pure.Actor
import adr.spine.pure.Admission
import adr.spine.pure.Context
import adr.spine.pure.ContextBounds
import adr.spine.pure.ContextFixture
import adr.spine.pure.DEFAULT_CONTEXT_BOUNDS
import adr.spine.pure.CURRENT_SCHEMA
import adr.spine.pure.Ctx
import adr.spine.pure.EffectKey
import adr.spine.pure.Fold
import adr.spine.pure.KeyedEffect
import adr.spine.pure.PerformMode
import adr.spine.pure.ProjectContext
import adr.spine.pure.Registry
import adr.spine.pure.SessionId
import adr.spine.pure.Signature
import adr.spine.pure.StagedInput
import adr.spine.pure.StepRecord
import adr.spine.pure.ContextRenderer
import adr.spine.ports.AuthorityResolver
import adr.spine.ports.Bus
import adr.spine.ports.Clock
import adr.spine.ports.ConfirmPolicy
import adr.spine.ports.IdSource
import adr.spine.ports.Sink

/**
 * Generic in the app's State type: that is the structural price of "the spine never
 * names a block" (§15 risk 6). The alternative — the spine importing blocks — breaks
 * G11 outright.
 */
class Boundary<S>(
    private val clock: Clock,
    private val ids: IdSource,
    private val bus: Bus,
    private val sink: Sink,
    private val authority: AuthorityResolver,
    private val policy: ConfirmPolicy,
    private val registry: Registry<S>,
    private val fold: Fold<S>,
    private val projectContext: ProjectContext<S>,
    private val promptVersion: String,
    private val session: SessionId,
    initial: S,
    /**
     * THE REASONER'S WINDOW, wired at the root (docs/DECISIONS.md:174), defaulted to
     * the spine's shipped one — the same shape [adr.spine.concurrency.SerialConsumer]
     * already gives its three deadlines.
     *
     * PUBLIC, and held HERE rather than curried into [projectContext], for one reason:
     * ONE value must reach both the digest this seam commits and the projection the
     * tools and the agent loop read. A root that curried the bound into the dispatcher
     * and declared another one here would have two, and nothing would notice.
     */
    val contextBounds: ContextBounds = DEFAULT_CONTEXT_BOUNDS,
) {
    var state: S = initial
        private set

    /**
     * Built here, not injected: the boundary must remain the SINGLE production site of
     * every ToolResult (gate check C7), which an outside-bound resolver would break.
     */
    private val actions = ActionResolution(registry)

    /** Built here for the same reason, and a stronger one: a bindable gate is bypassable. */
    private val irreversibility = IrreversibilityGate(registry, policy)

    /**
     * Derived from the SAME registry the gate reads (docs/DECISIONS.md:85), so admission
     * and the gate cannot disagree about which verbs are irreversible. The replay
     * harness is handed one built from that same registry, which is what makes
     * live == REPLAY == RECOVERY a property of the data rather than of two tables.
     */
    private val admission = Admission(registry)

    /** The bounded projection the reasoner sees right now — the loop stages it per step. */
    fun context(staged: List<StagedInput> = emptyList()): Context =
        projectContext(state, staged, contextBounds)

    /**
     * THE THREE CHANNELS, AND THEY ARE THE WHOLE PUBLIC STEP SURFACE. There is no
     * `onStepFinish` any more, because one entry taking the Actor as an argument is
     * one entry that lets its caller pick a principal — and `authorityOf` is asked
     * about exactly that value.
     *
     * Each is handed to one owner at wiring, and the Actor it stamps is fixed HERE,
     * in the only folder allowed to mint a `Signature` at all. §5.3's "decided by
     * where it entered, never by what it asks for" stops being a convention and
     * becomes the shape of a type: the payload has no property to ask with.
     */
    val human: Submit = Submit { commit(Actor.Human, it) }
    val agent: Submit = Submit { commit(Actor.Agent, it) }
    val spine: Submit = Submit { commit(Actor.Spine, it) }

    /**
     * PRIVATE, and that is the closure. `by` is a parameter of the CHANNEL and never
     * of the payload, so the only values it takes are the three above.
     */
    private fun commit(by: Actor, step: FinishedStep) {
        // 1 — the ONLY clock read in the system (G9).
        val now = clock.now()

        // 2 — the THIRD pure projection (G15). The tools see exactly what the reasoner saw.
        val ctx = Ctx(state, projectContext(state, step.staged, contextBounds))

        // 3 — the ONE closed name→ToolResult map (G1), before anything is stamped.
        val results = step.actions.map { actions.resolve(it, ctx) }

        // 4 — stamp WHO acted and resolve UNDER WHOSE PERMISSION, together, once (G1, G6).
        //     `by` came from the CHANNEL, not from `step`, so no caller decides which
        //     principal the authority resolver is asked about.
        val sig = Signature(by = by, authority = authority.authorityOf(by, session))

        // 5 — the gate, PRE-FOLD, keyed on the authority (G1/G6).
        val gated = results.map { irreversibility.check(it, sig, state) }

        // 6 — the pure decision. The only decider in the system.
        val (next, effects) = fold(state, gated, now, sig)

        // 7 — COMMIT the step as a unit (14.6). `results` is POST-GATE: exactly what was
        //     folded. `actions` is what was ASKED. 6.8: EVERY verb signs, presentation included.
        //     14.7: the envelope is stamped HERE, at the one site that mints a record,
        //     so no committed step can be missing its version.
        val index = bus.append(
            StepRecord(
                schemaVersion = CURRENT_SCHEMA,
                now = now,
                sig = sig,
                staged = step.staged,
                actions = step.actions,
                results = gated,
                commands = gated.map { actions.sign(it, sig, ids.next()) },
                context = ContextFixture(promptVersion, ContextRenderer().render(ctx.context)),
            ),
        )

        // 8 — adopt the derived cache.
        state = next

        // 9 — ADMIT, then perform with the key derived from the COMMITTED index (G9).
        //     This line literally cannot run before step 7, because `index` does not
        //     exist until then. The list handed to `perform` is FLAT: admission
        //     SUBSTITUTES a diagnostic in place rather than dropping, so the
        //     (step, index) key derivation is untouched.
        admission.admit(effects).forEachIndexed { i, effect ->
            sink.perform(KeyedEffect(EffectKey(index, i), effect), PerformMode.LIVE)
        }
    }
}

File: agent-driven-architecture-main/examples/typescript/src/spine/boundary/boundary.ts
// ── spine/boundary/boundary — THE ONE IMPURE SEAM (6.7, G9) ────────────────
// Clock, ids, bus, sink, authority, registry and the fold meet here and nowhere
// else. Nine ordered steps; three structural facts fall out of the order, and
// every implementation must preserve all three.
//
//  * COMMIT STRICTLY PRECEDES PERFORM — not by convention but because step 9
//    cannot run until step 7 has returned the StepIndex the key is built from.
//    14.6's ordering claim becomes unwritable-wrong.
//  * THE GATE RUNS BEFORE THE FOLD AND BEFORE THE COMMIT, so what is committed
//    is already the gate's verdict. A re-fold reproduces it without calling the
//    authorization seam again (G9).
//  * NOTHING DOWNSTREAM OF STEP 4 CAN LEARN WHO ACTED EXCEPT THROUGH `sig`. The
//    results were produced in step 3, before the signature existed. G1's
//    two-unreconciled-actor-values problem cannot recur, because there is only
//    one value and it is created after the tool has returned.
//
// AND NOTHING UPSTREAM OF STEP 4 CAN CHOOSE IT EITHER. `by` is a parameter of the
// CHANNEL a caller was handed, never a field of the step it submits, so the value
// fed to `authorityOf` is decided in this file and nowhere else.

import type { Authorization } from "../ports/authorization";
import type { Bus } from "../ports/bus";
import type { Clock } from "../ports/clock";
import type { IdSource } from "../ports/id-source";
import type { Sink } from "../ports/sink";
import type { Actor } from "../pure/actor";
import { Signature } from "../pure/actor";
import type { ContextBounds } from "../pure/context";
import { DEFAULT_CONTEXT_BOUNDS, render } from "../pure/context";
import type { Licences } from "../pure/effect";
import { admit, licencesOf } from "../pure/effect";
import type { SessionId, StepIndex } from "../pure/ids";
import { keyedEffect } from "../pure/keyed-effect";
import { SCHEMA_VERSION, type StepRecord } from "../pure/step-record";
import type { Ctx, Dispatchers } from "../pure/verb";
import type { FinishedStep, Registry, StepChannel } from "./action";
import { resolveAction, signResult } from "./action";
import { gate } from "./gate";

export interface BoundaryDeps<S> extends Dispatchers<S> {
  readonly clock: Clock;
  readonly ids: IdSource;
  readonly bus: Bus;
  readonly sink: Sink;
  readonly authz: Authorization<S>;
  readonly registry: Registry<S>;
  readonly session: SessionId;
  /** an injected asset (7.3, 14.7), captured on every committed record */
  readonly promptVersion: string;
  /** THE REASONER'S GROWTH BOUND, wired at the root (docs/DECISIONS.md:174).
   *  Omit it and the spine's shipped defaults apply — the same shape the
   *  mailbox deadlines already ship (spine/concurrency/consumer). It lives HERE
   *  rather than baked into `projectContext` so that ONE value reaches both the
   *  digest this seam commits and the projection the tools read, and a replay
   *  can be re-derived under a different one. */
  readonly contextBounds?: ContextBounds;
}

export class Boundary<S> {
  private current: S;

  /** DERIVED FROM THE REGISTRY THE GATE ALREADY READ, so admission and the gate
   *  cannot disagree about which verbs are irreversible (docs/DECISIONS.md:85).
   *  The same value is published by the root and handed to the replay harness,
   *  which is what makes live == REPLAY == RECOVERY a property of the data
   *  rather than of two independently-maintained tables. */
  private readonly licences: Licences;

  /** RESOLVED ONCE, for the reason `licences` is: a value re-defaulted per call
   *  is a value two call sites can disagree about, and the whole point of
   *  docs/DECISIONS.md:174 is that the bound the model saw and the bound the
   *  digest was derived under are one fact. */
  private readonly bounds: ContextBounds;

  /** THE THREE CHANNELS, AND THEY ARE THE WHOLE PUBLIC STEP SURFACE. There is no
   *  `onStepFinish` any more, because one entry taking the Actor as an argument
   *  is one entry that lets its caller pick a principal — and `authorityOf` is
   *  asked about exactly that value.
   *
   *  Each is handed to one owner at wiring, and the Actor it stamps is fixed
   *  HERE, in the only folder allowed to mint a `Signature` at all. §5.3's
   *  "decided by where it entered, never by what it asks for" stops being a
   *  convention and becomes the shape of a type: the payload has no field to ask
   *  with. */
  readonly human: StepChannel;
  readonly agent: StepChannel;
  readonly spine: StepChannel;

  constructor(
    private readonly deps: BoundaryDeps<S>,
    initial: S,
  ) {
    this.current = initial;
    this.licences = licencesOf(deps.registry.values());
    this.bounds = deps.contextBounds ?? DEFAULT_CONTEXT_BOUNDS;
    this.human = { submit: (step) => this.commit("Human", step) };
    this.agent = { submit: (step) => this.commit("Agent", step) };
    this.spine = { submit: (step) => this.commit("Spine", step) };
  }

  get state(): S {
    return this.current;
  }

  /** The wired bound, published so the agent loop and a replay site READ it
   *  instead of re-defaulting their own copy. */
  get contextBounds(): ContextBounds {
    return this.bounds;
  }

  /** PRIVATE, and that is the closure. `by` is a parameter of the CHANNEL and
   *  never of the payload, so the only values it takes are the three literals
   *  the constructor writes. */
  private commit(by: Actor, step: FinishedStep): StepIndex {
    // 1  the ONLY clock read in the system (G9)
    const now = this.deps.clock.now();

    // 2  the THIRD pure projection (G15) — the same Context the reasoner saw
    const ctx: Ctx<S> = {
      state: this.current,
      context: this.deps.projectContext(this.current, step.staged, this.bounds),
    };

    // 3  the ONE closed name→ToolResult map (G1)
    const results = step.actions.map((action) => resolveAction(this.deps.registry, action, ctx));

    // 4  stamp AND resolve authority (G1 + G6) — one value, created here, ever.
    //    `by` came from the CHANNEL, not from `step`, so no caller decides which
    //    principal the authorization seam is asked about.
    const sig = new Signature(by, this.deps.authz.authorityOf(by, this.deps.session));

    // 5  PRE-FOLD gate (G1/G6)
    const gated = results.map((r) =>
      gate(r, sig, this.current, this.deps.registry, this.deps.authz),
    );

    // 6  the pure decision — the only decider in the system
    const folded = this.deps.fold(this.current, gated, now, sig);

    // 7  COMMIT (14.6) — the step is the unit, and `now` rides it (G9).
    //    6.8: EVERY verb signs, presentation and domain alike.
    //    14.7: the envelope is stamped HERE, at the one site that mints a
    //    record, so no committed step can be missing its version.
    const record: StepRecord = {
      schemaVersion: SCHEMA_VERSION,
      now,
      sig,
      staged: step.staged,
      actions: step.actions,
      results: gated,
      commands: gated.map((r) => signResult(this.deps.registry, r, sig, this.deps.ids.next())),
      context: { promptVersion: this.deps.promptVersion, digest: render(ctx.context) },
    };
    const index = this.deps.bus.append(record);

    // 8  adopt the derived cache
    this.current = folded.state;

    // 9  ADMIT, then key from the COMMITTED index (G9) — the index is
    //    unavailable until step 7 returned, and the list handed to `perform` is
    //    FLAT: admission SUBSTITUTES a diagnostic in place rather than dropping,
    //    so the (step, index) key derivation is untouched.
    admit(this.licences, folded.effects).forEach((effect, i) => {
      this.deps.sink.perform(keyedEffect(index, i, effect), "LIVE");
    });

    return index;
  }
}
